spock-core/src/main/java/org/spockframework/runtime/extension/builtin/TagExtension.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.spockframework.runtime.extension.builtin;
+
+import org.spockframework.runtime.extension.IAnnotationDrivenExtension;
+import org.spockframework.runtime.model.FeatureInfo;
+import org.spockframework.runtime.model.SpecElementInfo;
+import org.spockframework.runtime.model.SpecInfo;
+import spock.lang.Tag;
+
+/**
+ * Processes {@link Tag @Tag} annotations on specifications and features, converting them to
+ * {@link org.spockframework.runtime.model.Tag Spock tags} which can potentially be used by reporting extensions.
+ * More importantly, it also makes the tag information available to JUnit 5 as
+ * {@link org.junit.platform.engine.TestTag platform tags} in order to enable filtering, e.g. from build tools like
+ * Maven or Gradle.
+ *
+ * @see <a href="https://junit.org/junit5/docs/current/user-guide/#running-tests-tag-expressions">
+ *   JUnit platform tag expression syntax</a>
+ * @author Alexander Kriegisch
+ * @since 2.2
+ */
+public class TagExtension implements IAnnotationDrivenExtension<Tag> {
+  public static final String TAG_EXTENSION_KEY = "TagExtension";
+
+  @Override
+  public void visitSpecAnnotation(Tag tag, SpecInfo spec) {
```

Comment:
Extensions with repeatable annotations should use the plural version `visitFeatureAnnotations` and `visitSpecAnnotations` instead.

Reply:
I do not understand. I have to implement the singular-type interface methods, while the plural-type ones have default implementations iterating over annotations and calling their respective singular-type counterparts.

Reply:
If you look at the `IAnnotationDrivenExtension` you see that all are `default` methods, and Spock will [only call the plural-type methods](https://github.com/spockframework/spock/blob/57e3c2b5a0b0fa6b7e2a5e007ca9b94433aa1643/spock-core/src/main/java/org/spockframework/runtime/ExtensionRunner.java#L123).

Reply:
Yes, but the ones I implemented are throwing exceptions for the default implementations. If there is a default implementation handling the iterations for me, why would I overwrite it only to re-implement the loop? That is nonsense IMHO. And relying on the fact that _currently_ Spock only calls the plural type methods, even though the singular ones exist, is a gamble I do not want to play. Without looking at the source code you pointed me to, I had no way of knowing it. Sacred knowledge in this case is not a good design guideline. I am not going to change it.